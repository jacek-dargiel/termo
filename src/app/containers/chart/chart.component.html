<div class="chart" *ngIf="location">
  <header class="header">
    <h3 class="header__location-name">🌡{{location?.name}}</h3>
    <div class="header__actions">
      <button
        type="button"
        class="btn btn--icon"
        (click)="close()"
      >️✖</button>
    </div>
  </header>
  <div class="chart__container">
    <ngx-charts-line-chart
      *ngIf="location && chartData"
      [results]="chartData"
      [xAxis]="true"
      [yAxis]="true"
      [roundDomains]="true"
      [xAxisTickFormatting]="formatTime"
    ></ngx-charts-line-chart>
  </div>
</div>
