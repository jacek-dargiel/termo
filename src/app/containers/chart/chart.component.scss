@use 'sass:color';
@use '../../../styles/variables' as vars;
@use '../../../../node_modules/@material/animation' as mdc;

$chartHeight: 15rem;

:host {
  position: fixed;
  bottom: 1.5rem;
  transform: translateY($chartHeight + 3rem);
  left: 1.5rem;
  right: 1.5rem;
  height: 15rem;
  background: #fff;
  border: 1px solid color.mix(#bbb, vars.$colorPrimary, 85%);
  box-shadow: 0.3rem 0.3rem 1rem rgba(#000, 0.5);
  transition: mdc.exit-temporary(transform, 175ms);
  display: flex;

  &.chart--visible {
    transition: mdc.enter(transform, 175ms);
    transform: translateY(0);
  }
}

.chart {
  flex: 1;
  margin: 0.75rem 1rem;
  display: flex;
  flex-direction: column;

  &__container {
    flex: 1;
    // width: 500px;
    // height: 300px;
  }
}
.header {
  display: flex;

  &__location-name {
    flex: 1;
    margin: 0;
    font-size: 1rem;
    text-transform: uppercase;
    color: color.mix(#777, vars.$colorPrimary, 80%);
  }

  &__actions {
    width: 2rem;

    .btn--icon {
      background: transparent;
      border: 0;
    }
  }
}
