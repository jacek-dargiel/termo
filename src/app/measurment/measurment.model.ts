export interface Measurment {
  id: string;
}
