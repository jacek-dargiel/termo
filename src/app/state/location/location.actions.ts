import { Action } from '@ngrx/store';
import { Location } from './location.model';

export enum LocationActionTypes {
  MapInitialized = '[Map] Map Initialized',
  RefreshMeasurmentsStart = '[Effect] Refresh Measurments Start',
  RefreshMeasurmentsFinish = '[Effect] Refresh Measurments Finish',
  FetchLocationsSuccess = '[API] Fetch Locations Success',
  FetchLocationsError = '[API] Fetch Locations Error',

  SelectLocation = '[Map] Select Location',
}

export class FetchLocationsSuccess implements Action {
  readonly type = LocationActionTypes.FetchLocationsSuccess;

  constructor(public payload: { locations: Location[] }) {}
}

// ### Custom actions

export class MapInitialized implements Action {
  readonly type = LocationActionTypes.MapInitialized;
}

export class RefreshMeasurmentsStart implements Action {
  readonly type = LocationActionTypes.RefreshMeasurmentsStart;
  constructor(public payload: { location: Location }) {}
}

export class RefreshMeasurmentsFinish implements Action {
  readonly type = LocationActionTypes.RefreshMeasurmentsFinish;
}

export class FetchLocationsError implements Action {
  readonly type = LocationActionTypes.FetchLocationsError;
  constructor(public payload: { error: Error }) {}
}

export class SelectLocation implements Action {
  readonly type = LocationActionTypes.SelectLocation;
  constructor(public payload: { location: Location }) {}
}

export type LocationActions =
 FetchLocationsSuccess
 | MapInitialized
 | RefreshMeasurmentsStart
 | RefreshMeasurmentsFinish
 | FetchLocationsError
 | SelectLocation
 ;
