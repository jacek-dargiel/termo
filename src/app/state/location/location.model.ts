import { Point } from '../../interfaces';

export interface Location {
  id: number;
  name: string;
  mapPosition: Point;
  feedKey: string;
  updatedAt: Date;
}
