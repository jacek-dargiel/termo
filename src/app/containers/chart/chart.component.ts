import { Component, OnInit, HostBinding, OnDestroy, inject } from '@angular/core';
import { ChartFacade } from './chart.facade';
import { Subscription } from 'rxjs';
import { Location } from '../../state/location/location.model';
import { map, filter } from 'rxjs/operators';
import { Measurment } from '../../state/measurment/measurment.model';
import { format } from 'date-fns';

@Component({
    selector: 'termo-chart',
    templateUrl: './chart.component.html',
    styleUrls: ['./chart.component.scss'],
    standalone: false
})
export class ChartComponent implements OnInit, OnDestroy {
  private chartFacade = inject(ChartFacade);

  @HostBinding('class.chart--visible') visible = false;
  selectedLocationSub: Subscription;
  locationMeasurmentsSub: Subscription;
  measurmentsSub: Subscription;
  location: Location;
  chartData;

  ngOnInit() {
    this.selectedLocationSub = this.chartFacade.selectedLocation$
      .subscribe(location => {
        this.location = location;
        this.visible = Boolean(location);
      });

    this.locationMeasurmentsSub = this.chartFacade.selectedLocationMeasurments$
      .pipe(
        filter(measurments => measurments !== undefined),
        map(measuments => this.mapMeasurmentToChartDataPoint(measuments))
      )
      .subscribe(data => {
        this.chartData = data;
      });
  }

  ngOnDestroy() {
    if (this.selectedLocationSub) {
      this.selectedLocationSub.unsubscribe();
    }
    if (this.locationMeasurmentsSub) {
      this.locationMeasurmentsSub.unsubscribe();
    }
  }

  close() {
    this.chartFacade.closeChart();
  }

  mapMeasurmentToChartDataPoint(measurments: Measurment[]) {
    let data = {
      name: this.location.name,
      series: measurments.map(measurment => ({
        name: measurment.created_at,
        value: measurment.value,
      })),
    };
    return [data];
  }

  formatTime(date: Date) {
    return format(date, 'HH:mm');
  }

}
