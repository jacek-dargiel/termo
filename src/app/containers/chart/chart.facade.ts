import { Injectable } from '@angular/core';
import { Store } from '@ngrx/store';
import { State } from '../../state/reducers';
import { selectSelectedLocation, selectSelectedLocationMeasurments } from '../../state/selectors';
import { SelectLocation } from '../../state/location/location.actions';

@Injectable({
  providedIn: 'root',
})
export class ChartFacade {
  public selectedLocation$ = this.store.select(selectSelectedLocation);
  public selectedLocationMeasurments$ = this.store.select(selectSelectedLocationMeasurments);

  constructor (
    private store: Store<State>
  ) {}

  closeChart() {
    this.store.dispatch(new SelectLocation({ location: undefined }));
  }
}
