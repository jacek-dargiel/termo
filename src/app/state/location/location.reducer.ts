import { EntityState, EntityAdapter, createEntityAdapter } from '@ngrx/entity';
import { Location } from './location.model';
import { LocationActions, LocationActionTypes } from './location.actions';
import { MeasurmentActionTypes, MeasurmentActions } from '../measurment/measurment.actions';
import { Measurment } from '../measurment/measurment.model';
import { compareAsc } from 'date-fns/esm';
import { Dictionary } from '@ngrx/entity/src/models';
import { last } from '../../helpers/lodash';

export interface State extends EntityState<Location> {
  loading: boolean;
  latestMeasurmentIDs: Dictionary<string>;
  selected: string;
}

function locationPositionComparer(a: Location, b: Location) {
  let aOffset = a.mapPosition.x + a.mapPosition.y;
  let bOffset = b.mapPosition.x + b.mapPosition.y;

  return aOffset - bOffset;
}

export let adapter: EntityAdapter<Location> = createEntityAdapter<Location>({
  sortComparer: locationPositionComparer,
});

export const INITIAL_STATE: State = adapter.getInitialState({
  loading: false,
  latestMeasurmentIDs: {},
  selected: undefined,
});

export function reducer(
  state = INITIAL_STATE,
  action: LocationActions | MeasurmentActions
): State {
  switch (action.type) {

    case LocationActionTypes.FetchLocationsSuccess: {
      let loadedState = {
        ...state,
      };
      return adapter.addAll(action.payload.locations, loadedState);
    }

    case LocationActionTypes.MapInitialized: {
      return {
        ...state,
        loading: true,
      };
    }

    case LocationActionTypes.RefreshButtonClick:
    case LocationActionTypes.FetchLocationsSuccess:
    case LocationActionTypes.RefreshSignal:
    {
      return {
        ...state,
        loading: true,
      };
    }

    case MeasurmentActionTypes.FetchMeasurmentsError: {
      return {
        ...state,
        loading: false,
      };
    }

    case LocationActionTypes.FetchLocationsError: {
      return {
        ...state,
        loading: false,
      };
    }

    case LocationActionTypes.RefreshMeasurmentsFinished: {
      return {
        ...state,
        loading: false,
      };
    }

    case MeasurmentActionTypes.FetchMeasurmentsSuccess: {
      let locationsState = state;
      if (action.payload.measurments.length > 0) {
        let sortedMeasurments = sortMeasurments(action.payload.measurments);
        let latestMeasurment = last(sortedMeasurments);
        locationsState = adapter.updateOne({id: action.payload.locationId, changes: { updatedAt: latestMeasurment.created_at }}, state);
        locationsState = {
          ...locationsState,
          latestMeasurmentIDs: {
            ...locationsState.latestMeasurmentIDs,
            [action.payload.locationId]: latestMeasurment.id,
          }
        };
      }

      return {
        ...locationsState,
      };
    }

    case LocationActionTypes.SelectLocation: {
      return {
        ...state,
        selected: action.payload.location && action.payload.location.id,
      };
    }

    default: {
      return state;
    }
  }
}

export const {
  selectIds,
  selectEntities,
  selectAll,
  selectTotal,
} = adapter.getSelectors();

function sortMeasurments(measurments: Measurment[]) {
  return [...measurments].sort((a: Measurment, b: Measurment) => compareAsc(a.created_at, b.created_at));
}
