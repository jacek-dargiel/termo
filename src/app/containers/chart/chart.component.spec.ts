import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';

import { ChartComponent } from './chart.component';
import { ChartFacade } from './chart.facade';
import { ReplaySubject } from 'rxjs';
import { Location } from '../../state/location/location.model';
import { Measurment } from '../../state/measurment/measurment.model';
import { LineChartModule } from '@swimlane/ngx-charts';

let mockLocation: Location = {
  id: 'temperatura.gorny_tunel',
  name: 'GÃ³rny Tunel',
  mapPosition: {
    x: 0.25,
    y: 0.75,
  },
  updatedAt: new Date('2018-09-19T23:45:00'),
};
let mockLocation2: Location = {
  id: 'temperatura.dolny_tunel',
  name: 'Dolny Tunel',
  mapPosition: {
    x: 0.5,
    y: 0.5,
  },
  updatedAt: new Date('2018-09-19T23:45:00'),
};
let mockMeasurmnets: Measurment[] = [
  {
    id: '123',
    value: 21.05,
    created_at: new Date('2018-09-19T23:47:00'),
    feed_id: 1234,
    feed_key: 'temperatura.gorny_tunel',
  },
  {
    id: '124',
    value: 20.33,
    created_at: new Date('2018-09-19T23:55:00'),
    feed_id: 1234,
    feed_key: 'temperatura.gorny_tunel',
  }
];

class MockChartFacade {
  public selectedLocation$ = new ReplaySubject<Location>();
  public selectedLocationMeasurments$ = new ReplaySubject<Measurment[]>();

  constructor() {
    this.selectedLocation$.next(mockLocation);
    this.selectedLocationMeasurments$.next(mockMeasurmnets);
  }

  public closeChart() {
    this.selectedLocation$.next(undefined);
  }
}

describe('ChartComponent', () => {
  let component: ChartComponent;
  let fixture: ComponentFixture<ChartComponent>;
  let facade: MockChartFacade;

  beforeEach(waitForAsync(() => {
    facade = new MockChartFacade();

    TestBed.configureTestingModule({
      imports: [ ChartComponent ],
      providers: [
        { provide: ChartFacade, useValue: facade }
      ],
      schemas: [NO_ERRORS_SCHEMA]
    })
    .overrideComponent(ChartComponent, { remove: { imports: [LineChartModule]}})
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(ChartComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });
  it('should render chart', () => {
    expect(fixture).toMatchSnapshot();
  });
  it('should hide chart when closed', () => {
    component.close();
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
  it('should render new location when location changes', () => {
    facade.selectedLocation$.next(mockLocation2);
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
  // it('should render new measurments when measurments change', () => {
  //   facade.selectedLocationMeasurments$.next(mockMeasurmnets2);
  //   fixture.detectChanges();
  //   let chart = fixture.debugElement.query(By.css('ngx-charts-line-chart'));
  // });
});
