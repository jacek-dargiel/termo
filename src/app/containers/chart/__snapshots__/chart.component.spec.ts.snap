// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartComponent should hide chart when closed 1`] = `
<termo-chart
  chartData={[Function Array]}
  chartFacade={[Function MockChartFacade]}
  location="undefined"
  locationMeasurmentsSub={[Function Subscriber]}
  selectedLocationSub={[Function Subscriber]}
  visible="false"
>
</termo-chart>
`;

exports[`ChartComponent should render chart 1`] = `
<termo-chart
  chartData={[Function Array]}
  chartFacade={[Function MockChartFacade]}
  location={[Function Object]}
  locationMeasurmentsSub={[Function Subscriber]}
  selectedLocationSub={[Function Subscriber]}
  visible={[Function Boolean]}
>
  <div
    class="chart"
  >
    <header
      class="header"
    >
      <h3
        class="header__location-name"
      >
        üå°G√≥rny Tunel
      </h3>
      <div
        class="header__actions"
      >
        <button
          class="btn btn--icon"
          type="button"
        >
          Ô∏è‚úñ
        </button>
      </div>
    </header>
    <div
      class="chart__container"
    >
      <ngx-charts-line-chart />
    </div>
  </div>
</termo-chart>
`;

exports[`ChartComponent should render new location when location changes 1`] = `
<termo-chart
  chartData={[Function Array]}
  chartFacade={[Function MockChartFacade]}
  location={[Function Object]}
  locationMeasurmentsSub={[Function Subscriber]}
  selectedLocationSub={[Function Subscriber]}
  visible={[Function Boolean]}
>
  <div
    class="chart"
  >
    <header
      class="header"
    >
      <h3
        class="header__location-name"
      >
        üå°Dolny Tunel
      </h3>
      <div
        class="header__actions"
      >
        <button
          class="btn btn--icon"
          type="button"
        >
          Ô∏è‚úñ
        </button>
      </div>
    </header>
    <div
      class="chart__container"
    >
      <ngx-charts-line-chart />
    </div>
  </div>
</termo-chart>
`;
